### ISSUE TITLE
Test upgrade from previous version to latest version in CI

### ISSUE THREAD

**Issue by kpollich** [elastic/elastic-package#1831]
We currently don't have automated test coverage in CI for integration upgrades. It could be helpful to run automated tests against the stack for each package that assert an upgrade from the previous published version to the latest version is successful. This would catch cases where both versions of a package are considered valid, but backwards compatibility is broken between versions. 

Relates to https://github.com/elastic/elastic-package/issues/579

**Comment by andrewkroh** [comment]
_I came here to open a new issue, but found this one already opened. Here is what I was going to write._

## Test package upgrades

There are bugs that could be detected if we could make elastic-package execute the package upgrade process. Roughly what I would like to see tested is this scenario:

1. Install the latest version from the EPR.
2. Do something to create indices. I think we need indices to catch mapping upgrade problems.
3. Install (upgrade) to package present in the pull request.
4. Run the system test.

I think this would help catch mapping and transform upgrade problems.

Related

- https://github.com/elastic/integrations/issues/10972 
- https://github.com/elastic/integrations/issues/10750
- https://github.com/elastic/integrations/issues/6554

**Comment by cmacknz** [comment]
+1, either as part of https://github.com/elastic/elastic-package/issues/579 or this issue we need this. We have shipped bugs this test would have caught.

**Comment by andrewkroh** [comment]
https://github.com/elastic/integrations/pull/13198 is a bug that likely would have been caught by upgrade testing. We didn't catch it because we didn't have existing data produced by the older version, but users did.

**Comment by mrodm** [comment]
@jsoriano 

> > In order to achieve this it would be needed to define a new type of tests as "upgrade" tests along with the current ones (system, pipeline, static...)::
> 
> Did you consider to make it a flag (or environment variable) for existing test runners? This would allow to have upgrade tests for each type of test, it could make sense to test upgrades in policy and system tests, and maybe in others.
> 
> > As these upgrade tests are based on the values given to the package policy, it should be added, for instance in integration packages, at least a test for each data stream. And probably, it should be added with different settings in the policy itself to provide a better coverage.
> 
> Yes, this is important, we cannot just make upgrade tests if we don't have a configuration. Because of this it would be specially useful to be able to reuse existing definitions for policy and system tests.

I didn't think about it, but it looks a good approach to follow.
If the same tests (for both system and policy tests) can be reused, it will help to start testing these upgrades.

Adding support for policy tests could be done as a separate issue :+1:

One concern of adding these upgrade tests to the system and policy tests is that it must change how it is setup the test.
Currently, the package to be tested is installed once at the beginning of the execution and `elastic-package` runs all the tests defined. Once finished all tests, the package is uninstalled.
Adding upgrade tests there, this process must change since for each test:
- the old version must be installed and configured in a package policy.
- waiting for documents
- the new version must be installed and configured in the same package policy.
- waiting for documents and run validations

Drawbacks of running these upgrade tests:
- All tests defined in the package must be run sequentially
    - All packages will take longer to finish their tests (run all the current system tests plus testing the upgrade).
    - This is going to affect packages like `network_traffic` or `zeek` that could last more than 2 hours to run all system tests sequentially (https://github.com/elastic/elastic-package/issues/787#issuecomment-2090971992) without considering the time spent in validating the upgrade process.
    - And other packages (`kafka` or `cloudflare_logpush`) that currently last more than 1hour because they are running tests sequentially and they could take advantage of running those tests in parallel. 

Would it be interesting to add a new configuration testing to disable the upgrade test in a given system/policy test?
Thinking that maybe not all system tests (or policy tests) are required to test the upgrade process.
It could be added a new configuration `upgrade_tests` as an object, in case it is required to add more settings in the future related to these type of tests. Example:
```yaml
upgrade_tests:
  enable: false
vars: ~
data_stream:
  vars:
    paths:
      - "{{SERVICE_LOGS_DIR}}/access.log*"
```


By default, should `elastic-package` run all the tests defined as upgrade tests too? Or keep them disabled by default?


> > When should we upgrade the package?
> 
> For system tests I think that it will be needed to wait for ingested documents, to ensure that the policy was actually working before the upgrade.
> 
> For policy tests we can directly update the policy after updating the package.

:+1:

> > TBD Wait for ingesting documents?
> > 
> > * It would involve start the service and wait for some documents (as in system tests)
> 
> We probably want this, at least we'd need to wait till the agent is effectively running the policy and this may be a way to detect it.
> 
> > TBD Wait for ingesting more documents?
> > 
> > * For some packages, it would involve maybe to re-start the service
> 
> Same thing, we would need this at least to ensure that the agent is running the policy. Though in this case we cannot rely only on this because the agent would be collecting documents already from the previous version.
> 
> This will be also relevant if we know of cases that would have been detected only by ingesting documents.

Agreed, I guess some errors will be just raised if there are documents ingested.

> Yep, we can split this in two phases. On the first one we do just the orchestration (installing old package, configure policy, install new package, upgrade policy), and on the second phase we would wait for documents before and after the upgrade.

Depending on how it is added this support, probably waiting for the ingestion of some documents could be tackled in the same issue. WDYT ?


> > Which validations should be performed after upgrading the package version?
> 
> We could perform the same validations as current tests. Wait for a given number of documents, check mappings and so on.

:+1:

> > Transform issues?
> 
> Let this apart because as you say there are more problems to solve in this area. Though we can run the same checks as we do now during system tests.

Ok, it could be kept the same validations performed right now in system tests.


> > Same variable definitions used for the current package can be used into the old version.
> 
> Yeah, we will need to do something about this, though maybe this is not needed in the first implementation as variables don't change so frequently. We will need this at some point because even if this doesn't happen so frequently, this is surely a source of upgrade problems that we want to test.

Ok, this could be left for a future change if needed.


> > Daily jobs in the integrations repository should run these upgrade tests too?
> 
> Good point. I think it is specially relevant to test in dailies because we would cover different stack versions, and because of this potentially different "old" versions.
> 
> > > * Those daily jobs run with the contents of the main branch, and therefore those versions are already published.
> 
> We would need to request all versions to the package registry, and use as old version the previous one to the current.
> 
> > > * It would be the same scenario as in the first question mentioned above.
> 
> Not so the same, because in this case the package would be exactly the same, and there would be no interest on testing this upgrade.

That's true, it is not the same scenario.

If we can achieve to get the previous version from EPR, it could be added to the daily jobs these tests too. However, this behavior is different from the one mentioned for Pull Requests (option `a)`), where it is chosen the latest version from EPR.

What about setting `--upgrade-from latest` in PRs (`latest` could be the default value) and `--upgrade-from previous` in daily jobs. That value could be set via environment variables in Buildkite definitions.
When testing locally, this flag could be used to test any specific version as the old one (e.g. `--upgrade-from 2.0.0`). Assuming there are no differences between package policy variables between those two versions.

In order to help disabling these tests if there is any issue, it will be also interesting to add a new flag `--no-upgrade-tests` and an environment variable to not run these new tests.
If that flag is added or the environment variable defined, `elastic-package` will run the tests the same way they are run now. 

---

Given these comments, this could be a draft of tasks for this issue:

Related to `elastic-package`:
- Add support to run upgrade tests as part of the system tests
    - Add orchestration logic to install and configure the old version of the package
        - Get the latest package version from EPR if `--upgrade-from latest`.
        - Get the previous package version from EPR if `--upgrade-from previous`.
    - Update system tests to ingest documents before and after the upgrade.
        - If it is difficult, this item could be moved to a separate issue when working on it.
    - Add information about failed tests in the xUnit reports
    - Add support to disable if required these upgrade tests globally (via `--no-upgrade-tests` flag and environment variable)
- Add support to run upgrade tests as part of the policy tests
    - Follow the same requirements as system tests.

Related to Integrations repo:
- If the default value for `--upgrade-from` flag is `latest`, it should not be changed anything for Pull Requests.
- Update the daily CI scripts to disable first running upgrade tests to avoid any issues.
    - This could be done before that new feature is part of a new release via environment variable, so tests can be enabled when required.
- Update the daily CI scripts to include the `--upgrade-from previous` variable.

**Comment by mrodm** [comment]
Talked offline with @jsoriano 

Take away from that discussion:
- decided to disable by default all these new tests.
    - if disabled, policy tests and system tests must run the same process.
        - package must be installed once at the beginning of the test suite, and it must be uninstalled at the end of test suite.
- it could be interesting to start with policy tests since they do not require to manage services, agents and wait for new documents.
- changes in CI pipelines will depend on how long it takes for these new tests to be completed
    - As these new tests require to install and uninstall the package for each test, executing all tests are going to be slower specially system tests.
    - This could affect packages with a lot of tests like aws, network_traffic or zeek among others.


> Drawbacks of running these upgrade tests:
> 
> * All tests defined in the package must be run sequentially
>   
>   * All packages will take longer to finish their tests (run all the current system tests plus testing the upgrade).
>   * This is going to affect packages like `network_traffic` or `zeek` that could last more than 2 hours to run all system tests sequentially ([Add ability to deploy custom elastic-agents on different OS or runtimes #787 (comment)](https://github.com/elastic/elastic-package/issues/787#issuecomment-2090971992)) without considering the time spent in validating the upgrade process.
>   * And other packages (`kafka` or `cloudflare_logpush`) that currently last more than 1hour because they are running tests sequentially and they could take advantage of running those tests in parallel.
